From: Philipp Zabel <p.zabel@pengutronix.de>
Date: Fri, 12 Feb 2016 13:38:04 +0100
Subject: [PATCH] drm/imx: ipuv3-plane: add support for YUV422, NV12, and NV16
 formats

Hook up support for DRM_FORMAT_YUV422, DRM_FORMAT_NV12,
and DRM_FORMAT_NV16.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>
---
 drivers/gpu/drm/imx/ipuv3-plane.c | 20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/imx/ipuv3-plane.c b/drivers/gpu/drm/imx/ipuv3-plane.c
index 3766e8785d3e..2e9214a0e881 100644
--- a/drivers/gpu/drm/imx/ipuv3-plane.c
+++ b/drivers/gpu/drm/imx/ipuv3-plane.c
@@ -42,6 +42,9 @@ static const uint32_t ipu_plane_formats[] = {
 	DRM_FORMAT_YVYU,
 	DRM_FORMAT_YUV420,
 	DRM_FORMAT_YVU420,
+	DRM_FORMAT_YUV422,
+	DRM_FORMAT_NV12,
+	DRM_FORMAT_NV16,
 	DRM_FORMAT_RGB565,
 	DRM_FORMAT_RGB565_A8,
 	DRM_FORMAT_BGR565_A8,
@@ -108,6 +111,7 @@ int ipu_plane_set_base(struct ipu_plane *ipu_plane, struct drm_framebuffer *fb,
 		ipu_plane->stride[1] = fb->pitches[1];
 		break;
 	case DRM_FORMAT_YUV420:
+	case DRM_FORMAT_YUV422:
 	case DRM_FORMAT_YVU420:
 		/*
 		 * Multiplanar formats have to meet the following restrictions:
@@ -135,6 +139,17 @@ int ipu_plane_set_base(struct ipu_plane *ipu_plane, struct drm_framebuffer *fb,
 				      fb->pitches[2] * y / 2 + x / 2 - eba;
 		ipu_plane->stride[1] = fb->pitches[1];
 		break;
+	case DRM_FORMAT_NV12:
+	case DRM_FORMAT_NV16:
+		if (cma_obj[0]->paddr != cma_obj[1]->paddr) {
+			DRM_DEBUG_KMS("plane base addresses differ.\n");
+			return -EINVAL;
+		}
+		ipu_plane->u_offset = cma_obj[1]->paddr + fb->offsets[1] +
+				      fb->pitches[1] * y / 2 + x / 2 - eba;
+		ipu_plane->v_offset = 0;
+		ipu_plane->stride[1] = fb->pitches[1];
+		break;
 	}
 
 	if (ipu_plane->enabled) {
@@ -318,7 +333,10 @@ int ipu_plane_mode_set(struct ipu_plane *ipu_plane, struct drm_crtc *crtc,
 	if (interlaced)
 		ipu_cpmem_interlaced_scan(ipu_plane->ipu_ch, fb->pitches[0]);
 
-	if (fb->pixel_format == DRM_FORMAT_YUV420) {
+	if (fb->pixel_format == DRM_FORMAT_YUV420 ||
+	    fb->pixel_format == DRM_FORMAT_YUV422 ||
+	    fb->pixel_format == DRM_FORMAT_NV12 ||
+	    fb->pixel_format == DRM_FORMAT_NV16) {
 		ipu_cpmem_set_yuv_planar_full(ipu_plane->ipu_ch,
 					      ipu_plane->stride[1],
 					      ipu_plane->u_offset,
